src/compiler/base/AwaitableNode.ts

```diff
@@ -0,0 +1,95 @@
+import * as ts from "typescript";
+import {Constructor} from "./../../Constructor";
+import * as errors from "./../../errors";
+import {AwaitableNodeStructure} from "./../../structures";
+import {callBaseFill} from "./../callBaseFill";
+import {insertIntoParent, removeChildrenWithFormatting, FormattingKind} from "./../../manipulation";
+import {Node} from "./../common";
+import {NamedNode} from "./../base";
+
+export type AwaitableNodeExtensionType = Node<ts.Node & { awaitModifier?: ts.AwaitKeywordToken; }>;
+
+export interface AwaitableNode {
+    /**
+     * If it's a generator function.
+     */
+    isAwait(): boolean;
+    /**
+     * Gets the await token or undefined if none exists.
+     */
+    getAwaitKeywordToken(): Node<ts.AwaitKeywordToken> | undefined;
+    /**
+     * Gets the await token or throws if none exists.
+     */
+    getAwaitKeywordTokenOrThrow(): Node<ts.AwaitKeywordToken>;
+    /**
+     * Sets if the node is a generator.
+     * @param value - If it should be a generator or not.
+     */
+    setIsAwait(value: boolean): this;
+}
+
+export function AwaitableNode<T extends Constructor<AwaitableNodeExtensionType>>(Base: T): Constructor<AwaitableNode> & T {
+    return class extends Base implements AwaitableNode {
+        isAwait() {
+            return this.compilerNode.awaitModifier != null;
+        }
+
+        getAwaitKeywordToken(): Node<ts.AwaitKeywordToken> | undefined {
+            const awaitModifier = this.compilerNode.awaitModifier;
+            return awaitModifier == null ? undefined : (this.global.compilerFactory.getNodeFromCompilerNode(awaitModifier, this.sourceFile) as Node<ts.AwaitKeywordToken>);
+        }
+
+        getAwaitKeywordTokenOrThrow(): Node<ts.AwaitKeywordToken> {
+            return errors.throwIfNullOrUndefined(this.getAwaitKeywordToken(), "Expected to find an await token.");
+        }
+
+        setIsAwait(value: boolean) {
+            const awaitModifier = this.getAwaitKeywordToken();
+            const isSet = awaitModifier != null;
+
+            if (isSet === value)
+                return this;
+
+            if (awaitModifier == null) {
+                const info = getAwaitInsertInfo(this);
+                insertIntoParent({
```

Comment:
I believe this will fail when there's no space between `for` and the parenthesis. ex. `for(` -> `for(await `.

Instead use `forKeyword.getEnd()` for the position and `newText: " await"`. Then use `removePrecedingSpaces: true` below (as shown).

Generally, I will try to add spaces at the beginning instead of the end for this reason and because the spaces preceding a node are included in their position (somewhat of a factor).

Reply:
Good catch!